src/Symfony/Component/Form/DependencyInjection/FormPass.php

```diff
@@ -80,16 +91,26 @@ public function process(ContainerBuilder $container)
             $typeExtensions[$extendedType][] = $serviceId;
         }
 
-        $definition->replaceArgument(2, $typeExtensions);
+        $allExtensions = array();
+        foreach ($typeExtensions as $extendedType => $extensions) {
+            $allExtensions[$extendedType] = new IteratorArgument(array_map(
+                function ($extensionId) {
+                    return new Reference($extensionId);
+                },
+                $extensions
+            ));
+        }
 
-        $guessers = array_keys($container->findTaggedServiceIds($this->formTypeGuesserTag));
-        foreach ($guessers as $serviceId) {
-            $serviceDefinition = $container->getDefinition($serviceId);
-            if (!$serviceDefinition->isPublic()) {
-                throw new InvalidArgumentException(sprintf('The service "%s" must be public as form type guessers are lazy-loaded.', $serviceId));
-            }
+        return $allExtensions;
+    }
+
+    private function processFormTypeGuessers(ContainerBuilder $container)
+    {
+        $guessers = array();
+        foreach (array_keys($container->findTaggedServiceIds($this->formTypeGuesserTag)) as $serviceId) {
```

Comment:
no need for array_keys, it only creates an intermediary array for no reason, let's do 
`foreach ($container->findTaggedServiceIds($this->formTypeGuesserTag) as $serviceId => $tags) {`